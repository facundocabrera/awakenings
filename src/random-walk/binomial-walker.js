import MersenneTwister from "mersenne-twister";
import { distribution, cumulative } from "../math/binomial";

const generator = new MersenneTwister();

/**
 * Binomial Walker.
 *
 * @param {Number} n independent Bernoulli trials
 * @param {Number} p successes probability
 */
function Binomial(n, p) {
  const d = distribution(n, p);
  const acc = cumulative(d);
  const len = acc.length;

  function walker() {
    const rand = generator.random();

    for (let i = 0; i < len; i++) {
      if (rand < acc[i]) {
        return i;
      }
    }

    return len;
  }

  return walker;
}

export { Binomial };
