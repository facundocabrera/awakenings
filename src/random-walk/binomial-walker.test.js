import sum from "lodash/sum";
import { Binomial } from "./binomial-walker";

test("random-walk/binomial", () => {
  const trials = 5;
  const success = 0.5;
  const occurences = new Array(trials).fill(0);

  const walker = Binomial(trials, success);

  const end = 1000;
  let counter = 0;

  while (counter < end) {
    occurences[walker()]++;
    counter++;
  }

  // Observar la campana que se genera en la distribucion de las ocurrencias
  // la joda viene por acÃ¡
  // console.log(occurences);

  expect(sum(occurences)).toBe(end);
  expect(occurences).toMatchSnapshot();
});
