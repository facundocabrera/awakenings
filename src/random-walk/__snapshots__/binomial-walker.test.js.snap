// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`random-walk/binomial 1`] = `
Array [
  29,
  154,
  315,
  304,
  198,
]
`;
